zld: Linking directly against `libc++.tbd` fails

### Zig Version

0.10.0-dev.4217+9d8cdb855

### Steps to Reproduce

```c++
// test.cpp
#include <new>
#include <cstdio>

int main() {
    int *x = new int;
    *x = 5;
    fprintf(stderr, "x: %d\n", *x);
    delete x;
}
```
```console
$ ./build/zig2 c++ test.cpp /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/lib/libc++.tbd -nostdlib++ -nostdinc++ -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1 -v
```

### Expected Behavior

Program should compile and link successfully.

It does work using `clang++` directly, or if you replace `zig ld` with `ld` in the failing linker invocation.

### Actual Behavior

```console
# zig c++ test.cpp /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/lib/libc++.tbd -nostdlib++ -nostdinc++ -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1 -v
Homebrew clang version 15.0.1
Target: x86_64-unknown-macosx12.6.0-unknown
Thread model: posix
InstalledDir: /usr/bin
 (in-process)
 "/Users/toadman667/repos/zig/build/zig2" -cc1 -triple x86_64-unknown-macosx12.6.0-unknown -Wundef-prefix=TARGET_OS_ -Werror=undef-prefix -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -mrelax-all --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name test.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=all -ffp-contract=on -fno-rounding-math -funwind-tables=2 -target-sdk-version=12.3 -fcompatibility-qualified-id-block-type-checking -fvisibility-inlines-hidden-static-local-var -target-cpu penryn -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debug-info-kind=standalone -dwarf-version=4 -debugger-tuning=lldb -target-linker-version 819.6 -v -fcoverage-compilation-dir=/Users/toadman667/repos/zig -nostdsysteminc -nobuiltininc -resource-dir /Users/toadman667/repos/zig/lib/clang/15.0.1 -dependency-file /Users/toadman667/.cache/zig/tmp/202d463f388980df-test.o.d -MT /Users/toadman667/.cache/zig/tmp/202d463f388980df-test.o -sys-header-deps -MV -isystem /Users/toadman667/repos/zig/lib/include -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk -iwithsysroot /usr/include -iwithsysroot /usr/local/include -iframeworkwithsysroot /System/Library/Frameworks -iframeworkwithsysroot /Library/Frameworks -D _DEBUG -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1 -stdlib=libc++ -O0 -fdeprecated-macro -fdebug-compilation-dir=/Users/toadman667/repos/zig -ferror-limit 19 -fsanitize=alignment,array-bounds,bool,builtin,enum,float-cast-overflow,function,integer-divide-by-zero,nonnull-attribute,null,pointer-overflow,return,returns-nonnull-attribute,shift-base,shift-exponent,signed-integer-overflow,unreachable,vla-bound -fsanitize-trap=alignment,array-bounds,bool,builtin,enum,float-cast-overflow,function,integer-divide-by-zero,nonnull-attribute,null,pointer-overflow,return,returns-nonnull-attribute,shift-base,shift-exponent,signed-integer-overflow,unreachable,vla-bound -stack-protector 2 -stack-protector-buffer-size 4 -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fmax-type-align=16 -fcolor-diagnostics -fno-spell-checking -target-cpu icelake-client -target-feature -16bit-mode -target-feature -32bit-mode -target-feature -3dnow -target-feature -3dnowa -target-feature +64bit -target-feature +adx -target-feature +aes -target-feature -amx-bf16 -target-feature -amx-int8 -target-feature -amx-tile -target-feature +avx -target-feature +avx2 -target-feature -avx512bf16 -target-feature +avx512bitalg -target-feature +avx512bw -target-feature +avx512cd -target-feature +avx512dq -target-feature -avx512er -target-feature +avx512f -target-feature -avx512fp16 -target-feature +avx512ifma -target-feature -avx512pf -target-feature +avx512vbmi -target-feature +avx512vbmi2 -target-feature +avx512vl -target-feature +avx512vnni -target-feature -avx512vp2intersect -target-feature +avx512vpopcntdq -target-feature -avxvnni -target-feature +bmi -target-feature +bmi2 -target-feature -branchfusion -target-feature -cldemote -target-feature +clflushopt -target-feature -clwb -target-feature -clzero -target-feature +cmov -target-feature +crc32 -target-feature +cx16 -target-feature +cx8 -target-feature -enqcmd -target-feature +ermsb -target-feature +f16c -target-feature -false-deps-getmant -target-feature -false-deps-lzcnt-tzcnt -target-feature -false-deps-mulc -target-feature -false-deps-mullq -target-feature -false-deps-perm -target-feature -false-deps-popcnt -target-feature -false-deps-range -target-feature -fast-11bytenop -target-feature +fast-15bytenop -target-feature -fast-7bytenop -target-feature -fast-bextr -target-feature +fast-gather -target-feature -fast-hops -target-feature -fast-lzcnt -target-feature -fast-movbe -target-feature +fast-scalar-fsqrt -target-feature -fast-scalar-shift-masks -target-feature +fast-shld-rotate -target-feature +fast-variable-crosslane-shuffle -target-feature +fast-variable-perlane-shuffle -target-feature +fast-vector-fsqrt -target-feature -fast-vector-shift-masks -target-feature +fma -target-feature -fma4 -target-feature +fsgsbase -target-feature +fsrm -target-feature +fxsr -target-feature +gfni -target-feature -harden-sls-ijmp -target-feature -harden-sls-ret -target-feature -hreset -target-feature -idivl-to-divb -target-feature +idivq-to-divl -target-feature +invpcid -target-feature -kl -target-feature -lea-sp -target-feature -lea-uses-ag -target-feature -lvi-cfi -target-feature -lvi-load-hardening -target-feature -lwp -target-feature +lzcnt -target-feature +macrofusion -target-feature +mmx -target-feature +movbe -target-feature -movdir64b -target-feature -movdiri -target-feature -mwaitx -target-feature +nopl -target-feature -pad-short-functions -target-feature +pclmul -target-feature -pconfig -target-feature -pku -target-feature +popcnt -target-feature -prefer-128-bit -target-feature +prefer-256-bit -target-feature -prefer-mask-registers -target-feature -prefetchwt1 -target-feature +prfchw -target-feature -ptwrite -target-feature +rdpid -target-feature -rdpru -target-feature +rdrnd -target-feature +rdseed -target-feature -retpoline -target-feature -retpoline-external-thunk -target-feature -retpoline-indirect-branches -target-feature -retpoline-indirect-calls -target-feature -rtm -target-feature +sahf -target-feature -sbb-dep-breaking -target-feature -serialize -target-feature -seses -target-feature +sgx -target-feature +sha -target-feature -shstk -target-feature +slow-3ops-lea -target-feature -slow-incdec -target-feature -slow-lea -target-feature -slow-pmaddwd -target-feature -slow-pmulld -target-feature -slow-shld -target-feature -slow-two-mem-ops -target-feature -slow-unaligned-mem-16 -target-feature -slow-unaligned-mem-32 -target-feature -soft-float -target-feature +sse -target-feature +sse2 -target-feature +sse3 -target-feature +sse4.1 -target-feature +sse4.2 -target-feature -sse4a -target-feature -sse-unaligned-mem -target-feature +ssse3 -target-feature -tagged-globals -target-feature -tbm -target-feature -tsxldtrk -target-feature -uintr -target-feature -use-glm-div-sqrt-costs -target-feature -use-slm-arith-costs -target-feature +vaes -target-feature +vpclmulqdq -target-feature +vzeroupper -target-feature -waitpkg -target-feature -wbnoinvd -target-feature -widekl -target-feature +x87 -target-feature -xop -target-feature +xsave -target-feature +xsavec -target-feature +xsaveopt -target-feature +xsaves -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /Users/toadman667/.cache/zig/tmp/202d463f388980df-test.o -x c++ test.cpp
clang -cc1 version 15.0.1 based upon LLVM 15.0.1 default target x86_64-apple-darwin21.6.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/Library/Frameworks"
ignoring duplicate directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include"
  as it is a non-system directory that duplicates a system directory
#include "..." search starts here:
#include <...> search starts here:
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/c++/v1
 /Users/toadman667/repos/zig/lib/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/System/Library/Frameworks (framework directory)
End of search list.
LLVM Emit Object... ar rcs /Users/toadman667/.cache/zig/o/3e7f3a615c18bcc28a97ce3d365c6500/libcompiler_rt.a /Users/toadman667/.cache/zig/o/3e7f3a615c18bcc28a97ce3d365c6500/libcompiler_rt.a.o
MachO Flush... zig ld -dynamic -syslibroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/lib/libc++.tbd /Users/toadman667/.cache/zig/o/f2fc9cca4fc33962af7d298684601f28/test.o /Users/toadman667/.cache/zig/o/3e7f3a615c18bcc28a97ce3d365c6500/libcompiler_rt.a -o a.out -lSystem -lc -L/usr/lib -L/usr/local/lib -F/System/Library/Frameworks -F/Library/Frameworks
error(link): undefined reference to symbol '__Znwm'
error(link):   first referenced in '/Users/toadman667/.cache/zig/o/f2fc9cca4fc33962af7d298684601f28/test.o'
error(link): undefined reference to symbol '__ZdlPv'
error(link):   first referenced in '/Users/toadman667/.cache/zig/o/f2fc9cca4fc33962af7d298684601f28/test.o'
thread 1148915 panic: attempt to unwrap error: UndefinedSymbolReference
/Users/toadman667/repos/zig/src/link/MachO/zld.zig:513:13: 0x10a8737d9 in link.MachO.zld.linkWithZld (zig2)
            return error.UndefinedSymbolReference;
            ^
/Users/toadman667/repos/zig/src/link/MachO.zig:438:22: 0x10a854c22 in link.MachO.flush (zig2)
        .one_shot => return zld.linkWithZld(self, comp, prog_node),
                     ^
/Users/toadman667/repos/zig/src/link.zig:656:23: 0x10a851905 in link.File.flush (zig2)
            .macho => return @fieldParentPtr(MachO, "base", base).flush(comp, prog_node),
                      ^
/Users/toadman667/repos/zig/src/Compilation.zig:2443:21: 0x10a7d9867 in Compilation.flush (zig2)
        else => |e| return e,
                    ^
/Users/toadman667/repos/zig/src/Compilation.zig:2421:9: 0x10a7cbba3 in Compilation.update (zig2)
        try comp.flush(main_progress_node);
        ^
/Users/toadman667/repos/zig/src/main.zig:3376:5: 0x10a75b74e in main.updateModule (zig2)
    try comp.update();
    ^
/Users/toadman667/repos/zig/src/main.zig:3063:21: 0x10a6a2c7e in main.buildOutputType (zig2)
        else => |e| return e,
                    ^
/Users/toadman667/repos/zig/src/main.zig:248:9: 0x10a64538c in main.mainArgs (zig2)
        return buildOutputType(gpa, arena, args, .cpp);
        ^
```

--- a/src/link/MachO/Dylib.zig
+++ b/src/link/MachO/Dylib.zig
@@ -22,7 +22,14 @@ weak: bool = false,
 /// Parsed symbol table represented as hash map of symbols'
 /// names. We can and should defer creating *Symbols until
 /// a symbol is referenced by an object file.
-symbols: std.StringArrayHashMapUnmanaged(void) = .{},
+///
+/// The value for each parsed symbol represents whether the
+/// symbol is defined as a weak symbol or strong.
+/// TODO when the referenced symbol is weak, ld64 marks it as
+/// N_REF_TO_WEAK but need to investigate if there's more to it
+/// such as weak binding entry or simply weak. For now, we generate
+/// standard bind or lazy bind.
+symbols: std.StringArrayHashMapUnmanaged(bool) = .{},
 
 pub const Id = struct {
     name: []const u8,
@@ -168,7 +175,7 @@ pub fn parseFromBinary(
                     if (!add_to_symtab) continue;
 
                     const sym_name = mem.sliceTo(@ptrCast([*:0]const u8, strtab.ptr + sym.n_strx), 0);
-                    try self.symbols.putNoClobber(allocator, try allocator.dupe(u8, sym_name), {});
+                    try self.symbols.putNoClobber(allocator, try allocator.dupe(u8, sym_name), false);
                 }
             },
             .ID_DYLIB => {
@@ -202,25 +209,30 @@ fn addObjCClassSymbol(self: *Dylib, allocator: Allocator, sym_name: []const u8) 
 
     for (expanded) |sym| {
         if (self.symbols.contains(sym)) continue;
-        try self.symbols.putNoClobber(allocator, sym, {});
+        try self.symbols.putNoClobber(allocator, sym, false);
     }
 }
 
 fn addObjCIVarSymbol(self: *Dylib, allocator: Allocator, sym_name: []const u8) !void {
     const expanded = try std.fmt.allocPrint(allocator, "_OBJC_IVAR_$_{s}", .{sym_name});
     if (self.symbols.contains(expanded)) return;
-    try self.symbols.putNoClobber(allocator, expanded, {});
+    try self.symbols.putNoClobber(allocator, expanded, false);
 }
 
 fn addObjCEhTypeSymbol(self: *Dylib, allocator: Allocator, sym_name: []const u8) !void {
     const expanded = try std.fmt.allocPrint(allocator, "_OBJC_EHTYPE_$_{s}", .{sym_name});
     if (self.symbols.contains(expanded)) return;
-    try self.symbols.putNoClobber(allocator, expanded, {});
+    try self.symbols.putNoClobber(allocator, expanded, false);
 }
 
 fn addSymbol(self: *Dylib, allocator: Allocator, sym_name: []const u8) !void {
     if (self.symbols.contains(sym_name)) return;
-    try self.symbols.putNoClobber(allocator, try allocator.dupe(u8, sym_name), {});
+    try self.symbols.putNoClobber(allocator, try allocator.dupe(u8, sym_name), false);
+}
+
+fn addWeakSymbol(self: *Dylib, allocator: Allocator, sym_name: []const u8) !void {
+    if (self.symbols.contains(sym_name)) return;
+    try self.symbols.putNoClobber(allocator, try allocator.dupe(u8, sym_name), true);
 }
 
 const TargetMatcher = struct {
@@ -359,6 +371,12 @@ pub fn parseFromStub(
                             }
                         }
 
+                        if (exp.weak_symbols) |symbols| {
+                            for (symbols) |sym_name| {
+                                try self.addWeakSymbol(allocator, sym_name);
+                            }
+                        }
+
                         if (exp.objc_classes) |objc_classes| {
                             for (objc_classes) |class_name| {
                                 try self.addObjCClassSymbol(allocator, class_name);
@@ -402,6 +420,12 @@ pub fn parseFromStub(
                             }
                         }
 
+                        if (exp.weak_symbols) |symbols| {
+                            for (symbols) |sym_name| {
+                                try self.addWeakSymbol(allocator, sym_name);
+                            }
+                        }
+
                         if (exp.objc_classes) |classes| {
                             for (classes) |sym_name| {
                                 try self.addObjCClassSymbol(allocator, sym_name);
@@ -429,6 +453,12 @@ pub fn parseFromStub(
                         if (reexp.symbols) |symbols| {
                             for (symbols) |sym_name| {
                                 try self.addSymbol(allocator, sym_name);
+                            }
+                        }
+
+                        if (reexp.weak_symbols) |symbols| {
+                            for (symbols) |sym_name| {
+                                try self.addWeakSymbol(allocator, sym_name);
                             }
                         }
 

--- a/src/link/tapi.zig
+++ b/src/link/tapi.zig
@@ -26,6 +26,7 @@ pub const TbdV3 = struct {
         allowable_clients: ?[]const []const u8,
         re_exports: ?[]const []const u8,
         symbols: ?[]const []const u8,
+        weak_symbols: ?[]const []const u8,
         objc_classes: ?[]const []const u8,
         objc_ivars: ?[]const []const u8,
         objc_eh_types: ?[]const []const u8,
@@ -53,6 +54,7 @@ pub const TbdV4 = struct {
     exports: ?[]const struct {
         targets: []const []const u8,
         symbols: ?[]const []const u8,
+        weak_symbols: ?[]const []const u8,
         objc_classes: ?[]const []const u8,
         objc_ivars: ?[]const []const u8,
         objc_eh_types: ?[]const []const u8,
@@ -60,6 +62,7 @@ pub const TbdV4 = struct {
     reexports: ?[]const struct {
         targets: []const []const u8,
         symbols: ?[]const []const u8,
+        weak_symbols: ?[]const []const u8,
         objc_classes: ?[]const []const u8,
         objc_ivars: ?[]const []const u8,
         objc_eh_types: ?[]const []const u8,

--- a/test/link.zig
+++ b/test/link.zig
@@ -79,6 +79,11 @@ fn addWasmCases(cases: *tests.StandaloneContext) void {
 }
 
 fn addMachOCases(cases: *tests.StandaloneContext) void {
+    cases.addBuildFile("test/link/macho/bugs/13056/build.zig", .{
+        .build_modes = true,
+        .requires_macos_sdk = true,
+    });
+
     cases.addBuildFile("test/link/macho/bugs/13457/build.zig", .{
         .build_modes = true,
     });

--- a/test/link/macho/bugs/13056/build.zig
+++ b/test/link/macho/bugs/13056/build.zig
@@ -1,0 +1,29 @@
+const std = @import("std");
+const Builder = std.build.Builder;
+
+pub fn build(b: *Builder) void {
+    const mode = b.standardReleaseOptions();
+
+    const target: std.zig.CrossTarget = .{ .os_tag = .macos };
+    const target_info = std.zig.system.NativeTargetInfo.detect(target) catch unreachable;
+    const sdk = std.zig.system.darwin.getDarwinSDK(b.allocator, target_info.target) orelse
+        @panic("macOS SDK is required to run the test");
+
+    const test_step = b.step("test", "Test the program");
+
+    const exe = b.addExecutable("test", null);
+    b.default_step.dependOn(&exe.step);
+    exe.addIncludePath(std.fs.path.join(b.allocator, &.{ sdk.path, "/usr/include" }) catch unreachable);
+    exe.addIncludePath(std.fs.path.join(b.allocator, &.{ sdk.path, "/usr/include/c++/v1" }) catch unreachable);
+    exe.addCSourceFile("test.cpp", &.{
+        "-nostdlib++",
+        "-nostdinc++",
+    });
+    exe.addObjectFile(std.fs.path.join(b.allocator, &.{ sdk.path, "/usr/lib/libc++.tbd" }) catch unreachable);
+    exe.setBuildMode(mode);
+
+    const run_cmd = exe.run();
+    run_cmd.expectStdErrEqual("x: 5\n");
+
+    test_step.dependOn(&run_cmd.step);
+}

--- a/test/link/macho/bugs/13056/test.cpp
+++ b/test/link/macho/bugs/13056/test.cpp
@@ -1,0 +1,10 @@
+// test.cpp
+#include <new>
+#include <cstdio>
+
+int main() {
+    int *x = new int;
+    *x = 5;
+    fprintf(stderr, "x: %d\n", *x);
+    delete x;
+}

